src/java.base/macosx/native/libjli/java_md_macosx.m

```diff
@@ -830,38 +830,53 @@ static void MacOSXStartup(int argc, char *argv[]) {
         (*env)->ExceptionClear(env);
         return;
     }
 
-    const char *mainClassName = NULL;
-    NULL_CHECK(mainClassName = (*env)->GetStringUTFChars(env, mainClassString, NULL));
-
-    char envVar[80];
-    /*
-     * The JAVA_MAIN_CLASS_<pid> environment variable is used to pass
-     * the name of a Java class whose main() method is invoked by
-     * the Java launcher code to start the application, to the AWT code
-     * in order to assign the name to the Apple menu bar when the app
-     * is active on the Mac.
-     * The _<pid> part is added to avoid collisions with child processes.
-     *
-     * WARNING: This environment variable is an implementation detail and
-     * isn't meant for use outside of the core platform. The mechanism for
-     * passing this information from Java launcher to other modules may
-     * change drastically between update release, and it may even be
-     * removed or replaced with another mechanism.
+    /* There are multiple apple.awt.*" system properties that AWT(the desktop module)
+     * references that are inherited from Apple JDK.
+     * This inherited AWT code looks for this property and uses it for the name
+     * of the app as it appears in the system menu bar.
      *
-     * NOTE: It is used by SWT, and JavaFX.
+     * Not idea if how much external code ever sets it, but use it if set, else
+     * if not set (the high probability event) set it to the application class name.
      */
-    snprintf(envVar, sizeof(envVar), "JAVA_MAIN_CLASS_%d", getpid());
-    setenv(envVar, mainClassName, 1);
+    const char* propName = "apple.awt.application.name";
+    jstring jKey = NULL;
+    NULL_CHECK(jKey = (*env)->NewStringUTF(env, propName));
+
+    jclass sysClass = NULL;
+    NULL_CHECK(sysClass = (*env)->FindClass(env, "java/lang/System"));
+
+    jmethodID getPropertyMID = NULL;
+    NULL_CHECK(getPropertyMID = (*env)->GetStaticMethodID(env, sysClass,
+               "getProperty", "(Ljava/lang/String;)Ljava/lang/String;"));
+
+    jmethodID setPropertyMID = NULL;
+    NULL_CHECK(setPropertyMID = (*env)->GetStaticMethodID(env, sysClass,
+               "setProperty",
+               "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;"));
+
+    jstring jValue = (*env)->CallStaticObjectMethod(env, sysClass, getPropertyMID, jKey);
+    if ((*env)->ExceptionCheck(env)) {
+        (*env)->ExceptionClear(env);
+        (*env)->DeleteLocalRef(env, jKey);
+        return;
+    }
+    if (jValue == NULL) {
+        (*env)->CallStaticObjectMethod(env, sysClass, setPropertyMID,
+                                       jKey, mainClassString);
+        if ((*env)->ExceptionCheck(env)) {
+            (*env)->ExceptionClear(env);
+            (*env)->DeleteLocalRef(env, jKey);
+            return;
+        }
+    } else {
+        (*env)->DeleteLocalRef(env, jValue);
+    }
 
-    (*env)->ReleaseStringUTFChars(env, mainClassString, mainClassName);
+    (*env)->DeleteLocalRef(env, jKey);
```

Comment:
I am not sure about error handling, the jkey is not removed when NULL_CHECK is used. Also an exceptions are not cleared in case of NULL_CHECK like in other cases, is it intentional?

Reply:
Well .. they aren't removed by the existing code either. And this is the launcher.
So far as I can tell if we error out of here (as I found when I made a typo in a method signature) the
VM exits very rapidly. So if I do anything here, it would be to remove DeleteLocalRef since it really doesn't matter to clean up local refs when you are either bailing from native .. or the entire process .. cleaning up when you are continuing on (as the code does) is perhaps more important since you might need more local refs before you are done.
Or file a separate bug against the launcher and JNI clean up after error handling.
java.base/share/native/libjli/java.c is a good example of where the same pattern exists.